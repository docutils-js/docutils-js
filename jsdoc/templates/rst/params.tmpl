<?js
    var params = obj;

    /* sort subparams under their parent params (like opts.classname) */
    var parentParam = null;
    params.forEach(function(param, i) {
        var paramRegExp;

        if (!param) {
            return;
        }

        if (parentParam && parentParam.name && param.name) {
            try {
                paramRegExp = new RegExp('^(?:' + parentParam.name + '(?:\\[\\])*)\\.(.+)$');
            }
            catch (e) {
                // there's probably a typo in the JSDoc comment that resulted in a weird
                // parameter name
                return;
            }

            if ( paramRegExp.test(param.name) ) {
                param.name = RegExp.$1;
                parentParam.subparams = parentParam.subparams || [];
                parentParam.subparams.push(param);
                params[i] = null;
            }
            else {
                parentParam = param;
            }
        }
        else {
            parentParam = param;
        }
    });

    /* determine if we need extra columns, "attributes" and "default" */
    params.hasAttributes = false;
    params.hasDefault = false;
    params.hasName = false;

    params.forEach(function(param) {
        if (!param) { return; }

        if (param.optional || param.nullable || param.variable) {
            params.hasAttributes = true;
        }

        if (param.name) {
            params.hasName = true;
        }

        if (typeof param.defaultvalue !== 'undefined') {
            params.hasDefault = true;
        }
    });
?>

.. csv-table::
:header: <?js= <?js if (params.hasName) {?>Name, <?js } ?>Type, <?js if (params.hasAttributes) {?>Attributes, <?js } ?><?js if (params.hasDefault) {?>Default, <?js } ?>Description
<?js
        var self = this;
        params.forEach(function(param) {
            if (!param) { return; }
    ?><?js if (params.hasName) {?>"`?js= param.name ?>`"<?js } ?><?js if (param.type && param.type.names) {?><?js= self.partial('type.tmpl', param.type.names) ?><?js } ?><?js if (params.hasAttributes) {?><?js if (param.optional) { ?>"<optional>"<?js } ?><?js if (param.nullable) { ?>"<nullable>",<?js } ?><?js if (param.variable) { ?>"<repeatable>",<?js } ?>
<?js } ?>
<?js if (params.hasDefault) {?>
                <td class="default">
                <?js if (typeof param.defaultvalue !== 'undefined') { ?>
                    <?js= self.htmlsafe(param.defaultvalue) ?>
                <?js } ?>
                </td>
            <?js } ?>

            <td class="description last"><?js= param.description ?><?js if (param.subparams) { ?>
                <h6>Properties</h6>
                <?js= self.partial('params.tmpl', param.subparams) ?>
            <?js } ?></td>
        </tr>

    <?js }); ?>
    </tbody>
</table>
