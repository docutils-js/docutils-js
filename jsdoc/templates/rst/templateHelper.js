const helper = require('jsdoc/util/templateHelper');

helper.fileExtension = '.txt';

Object.keys(helper).forEach(k => exports[k] = helper[k]);
