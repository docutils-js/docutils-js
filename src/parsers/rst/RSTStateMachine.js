import { StateMachineWS } from '../../StateMachine';
import Inliner from './Inliner';
import * as languages from '../../languages';


class RSTStateMachine extends StateMachineWS {
    run({
 inputLines, document, inputOffset, matchTitles, inliner,
}) {
        if(inputOffset === undefined) {
            inputOffset = 0;
        }
        if (!document) {
            throw new Error('need document');
        }

        if (matchTitles === undefined) {
            matchTitles = true;
        }
        this.language = languages.getLanguage(document.settings.languageCode);
        this.matchTitles = matchTitles;
        if (inliner === undefined) {
            inliner = new Inliner();
        }
        inliner.initCustomizations(document.settings);
        this.memo = {
	    document,
	    reporter: document.reporter,
	    language: this.language,
            titleStyles: [],
                      sectionLevel: 0,
                      sectionBubbleUpKludge: false,
                      inliner,
};
        this.document = document;
        this.attachObserver(document.noteSource.bind(document));
        this.reporter = this.memo.reporter;
        this.node = document;
        const results = super.run({ inputLines, inputOffset, inputSource: document.source });
        if (results.length !== 0) {
//          throw new Error("should be o");
        }
        this.node = this.memo = undefined;
    }
}

export default RSTStateMachine;
