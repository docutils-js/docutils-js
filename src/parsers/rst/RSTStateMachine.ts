import StateMachineWS from '../../StateMachineWS';
import Inliner from './Inliner';
import * as languages from '../../languages';
import {ElementInterface, StateMachineRunArgs} from "../../types";
import { RSTLanguage, RstMemo } from "./types";
import { getLanguage } from "./languages";

/**
 * reStructuredText's master StateMachine.
 *
 * The entry point to reStructuredText parsing is the `run()` method.
 */
class RSTStateMachine extends StateMachineWS {
    public rstLanguage?: RSTLanguage;
    public matchTitles?: boolean;
    public node?: ElementInterface;
    public memo?: RstMemo;
    public run(args: StateMachineRunArgs): (string|{})[] {
        const cArgs = { ... args };
        /* istanbul ignore if */
        if (cArgs.inputOffset === undefined) {
            cArgs.inputOffset = 0;
        }
        /* istanbul ignore if */
        if (!cArgs.document) {
            throw new Error('need document');
        }

        const document = cArgs.document;
        /* istanbul ignore next */
        if (cArgs.matchTitles === undefined) {
            cArgs.matchTitles = true;
        }
        const languageCode = document.settings.docutilsCoreOptionParser.languageCode;
        if(languageCode !== undefined) {
            this.rstLanguage = getLanguage(languageCode);
        }
        this.matchTitles = cArgs.matchTitles;
        /* istanbul ignore next */
        if (cArgs.inliner === undefined) {
            cArgs.inliner = new Inliner(document);
        }
        cArgs.inliner.initCustomizations(document.settings);
        this.memo = {
            document,
            reporter: document.reporter,
            language: this.rstLanguage,
            titleStyles: [],
            sectionLevel: 0,
            sectionBubbleUpKludge: false,
            inliner: cArgs.inliner,
        };
        this.document = document;
        this.attachObserver(document.noteSource.bind(document));
        this.reporter = this.memo.reporter;
        this.node = document;
        const superArgs: StateMachineRunArgs =  {inputLines: cArgs.inputLines, inputOffset: cArgs.inputOffset, inputSource: this.document.source};
        const results = super.run(superArgs);

        /* istanbul ignore if */
        if (results.length !== 0) {
            throw new Error('should be empty array return from statemachine.run');
        }
        this.node = undefined;
        this.memo = undefined;

        return [];
    }
}

export default RSTStateMachine;
