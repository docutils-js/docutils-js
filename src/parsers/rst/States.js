import { StateMachineWS, StateWS } from '../../StateMachine';

export class RSTStateMachine extends StateMachineWS {
    run({inputLines, document, inputOffset, matchTitles, inliner}) {
	if(matchTitles === undefined) {
	    matchTitles = True;
	}
	this.language = languages.getLanguage(document.settings.languageCode)
	self.matchTitles = matchTitles;
	if(inliner === undefined) {
	    inliner = new Inliner();
	}
	inliner.initCustomizations(document.settings);
	this.memo = { document, reporter: document.reporter, language: this.language,
		      titleStyles: [],
		      sectionLevel: 0,
		      sectionBubbleUpKludge: false,
		      inliner };
	this.document = document;
	this.attachObserver(document.noteSource);
	this.reporter = this.memo.reporter
	this.node = document
	results = super.run({inputLines, inputOffset, inputSource: document.source});
	self.node = self.memo = undefined;
    }
}

class NestedStateMachine extends StateMachineWS {
    run({inputLines, inputOffset, memo, node, matchTitles}) {
	if(matchTitles === undefined) {
	    matchTitles = true;
	}
	this.matchTitles = matchTitles;
	this.memo = memo;
	this.document = memo.document
	this.attachObserver(this.document.noteSource);
	this.reporter = memo.reporter;
	this.language = memo.language;
	this.node = node;
	const results = super.run({inputLines, inputOffset});
	return results;
    }


}

class RSTState extends StateWS {
    constructor(stateMachine, debug) {
	super(stateMachine, debug);
	this.nestedSmKwargs = { stateClasses, initialState: 'Body' };
    }

    runtimeInit() {
	super.runtimeInit();
	const memo = this.stateMachine.me2mo;
	this.memo = memo;
	this.reporter = memo.reporter;
	this.inliner = memo.inliner;
	this.document = memo.document;
	this.parent = this.stateMachine.node;
	if(!this.reporter.getSourceAndLine) {
	    this.reporter.getSourceAndLine = this.stateMachine.getSourceAndLine;
	}

    }

    gotoLine(absLineOffset) {
	try {
	    this.stateMachine.gotoLine(absLineOffset);
	} catch(ex) {
	    /* test for eof error? */
	}
    }

    noMatch(context, transitins) {
	this.reporter.severe(`Internal error: no transition pattern match.  State: "${this.__class__.__name__}"; transitions: ${transitions}; context: ${context}; current line: ${this.stateMachine.line}.`);
	return [ context, null, [] ];
    }

    bof(context) {
	return [[], []]
    }

    nestedParse({ block, inputOffset, node, matchTitles, stateMachineClass, stateMachineKwargs}) {
	let useDefault = 0;
	if(!stateMachineClass) {
	    stateMachineClass = this.nestedSm;
	    useDefault = useDefault + 1;
	}
	if(!stateMachineKwargs) {
	    stateMachineKwargs = this.nestedSmKwargs;
	    useDefault = useDefault + 1;
	}
	let blockLength = block.length;

	let stateMachine = undefined;
	if(useDefault === 2) {
	    try {
		stateMachine = this.nestedSmCache.pop();
	    } catch(err) {
	    }
	}
	
	if(!stateMachine) {
	    stateMachine = stateMachineClass({debug:this.debug,
					      ...stateMachineKwags});
	}
	stateMachine.run({block, inputOffset, memo: this.memo,
			  node, matchTitles});
	if(useDefault === 2) {
	    this.nestedSmCache.push(stateMachine);
	} else {
	    stateMachine.unlink();
	}
	newOffset = stateMachine.absLineOffset();
	if(block.parent && (len(block) - block_length) !== 0) {
	    this.stateMachine.nextLine(block.length - blockLength);
	}
	return newOffset;
    }

    nestedListParse({block, inputOffset, node, initialState,
		     blankFinish, blankFinishState, extraSettings,
		     matchTitles,
		     stateMachineClass,
		     stateMachineKwargs}) {
	if(!stateMachineClass) {
	    stateMachineClass = this.nestedSm;
	}
	if(!stateMachineKwargs) {
	    state_machine_kwargs = { ... this.stateMachineKwargs };
	}
	stateMachineKwargs.initialState = initialState;
	const stateMachine = new stateMachineClass({debug: this.debug,
						    ... stateMachineKwargs});
	if(!blackFinishState) {
	    blankFinishState = initialState;
	}
	stateMachine.states[blankFinishState].blankFinish = blankFinish;
	Object.keys(extraSettings).forEach(key => {
	    stateMachine.states[initialState][key] = extraSettings[key];
	});
	stateMachine.run({block, inputOffset, memo: this.memo,
			  node, matchTitles});
	blankFinish = stateMachine.states[blankFinishState].blankFinish;
	stateMachine.unlink();
	return [stateMachine.absLineOffset(), blank_finish]
    }

    section({title, source, style, lineno, messages}) {
	if(this.checkSubsection({source, style, lineno})) {
	    this.newSubsection({title, lineno, messages});
	}
    }
}

	
					    
					     
		
    
	
    
