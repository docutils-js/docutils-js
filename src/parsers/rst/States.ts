import * as statemachine from '../../StateMachine';
import * as nodes from '../../nodes';

/* import RSTStateMachine from './RSTStateMachine';
import Inliner from './Inliner'; */

/* eslint-disable-next-line no-unused-vars */
const nonWhitespaceBefore = '(?<!\\s)';
/* eslint-disable-next-line no-unused-vars */
const nonWhitespaceEscapeBefore = '(?<![\\s\\x00])';
/* eslint-disable-next-line no-unused-vars */
const nonUnescapedWhitespaceEscapeBefore = '(?<!(?<!\\x00)[\\s\\x00])';
/* eslint-disable-next-line no-unused-vars */
const nonWhitespaceAfter = '(?!\\s)';
/* eslint-disable-next-line no-unused-vars */
const classifierDelimiterRegexp = new RegExp(' +: +');
/* eslint-disable-next-line no-unused-vars */
const simpleTableBorderPat = /=+[ =]*$/;
/* eslint-disable-next-line no-unused-vars */
const gridTableTopPat = /\+-[-+]+-\+ *$/;
/* eslint-disable-next-line no-unused-vars */
const emailPattern = '%(emailc)s+(?:\\.%(emailc)s+)*(?<!\x00)@%(emailc)s+(?:\\.%(emailc)s*)*%(uri_end)s';
// email=re.compile(self.email_pattern % args + '$',
//                 re.VERBOSE | re.UNICODE),

/* eslint-disable-next-line no-unused-vars */
const { StringList } = statemachine;

/* eslint-disable-next-line no-unused-vars */
const nonalphanum7bit = '[!-/:-@[-`{-~]';
/* eslint-disable-next-line no-unused-vars */
const simplename = '\\w+';

/* eslint-disable-next-line no-unused-vars */
const normalizeName = nodes.fullyNormalizeName;

/* eslint-disable-next-line no-unused-vars */
const { StateMachineWS } = statemachine;
/* eslint-disable-next-line no-unused-vars */
const { StateWS } = statemachine;


// SubstitutionDef];

/*    underline(match, context, nextState) {
        const overline = context[0]
        const blocktext = overline + '\n' + this.stateMachine.line
        const lineno = this.stateMachine.absLineNumber() - 1
//        if len(overline.rstrip()) < 4:
//            this.shortOverline(context, blocktext, lineno, 1)
        const msg = this.reporter.error(
              'Invalid section title or transition marker.',
            [new nodes.literal_block(blocktext, blocktext)],
            { line: lineno })
        this.parent.add(msg)
        return [[], 'Body', []]
    }
*/
