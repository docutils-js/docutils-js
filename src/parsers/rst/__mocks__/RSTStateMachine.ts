import StringList from '../../../StringList';
export const mockGetSourceAndLine = jest.fn(() => (['', 0]));
export const mockGetFirstKnownIndented = jest.fn((args) => ([new StringList(['hello']), args.indent, 0, true]));

const mock = jest.fn().mockImplementation(() => {
    return { getSourceAndLine: mockGetSourceAndLine,
        getFirstKnownIndented: mockGetFirstKnownIndented,
    };
});
export default mock;
