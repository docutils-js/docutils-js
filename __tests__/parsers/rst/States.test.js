import { Body, stateClasses } from '../../../src/parsers/rst/States';
import RSTStateMachine from '../../../src/parsers/rst/RSTStateMachine';
jest.mock('../../../src/parsers/rst/RSTStateMachine');

function createsm() {
        const sm = new RSTStateMachine({ stateClasses, initialState: 'Body',
				     debug: true,
					 debugFn: console.log });
    return sm;
}
test.only('Body patterns', () => {
    const stateMachine = createsm();
    const body = new Body({ stateMachine });
    expect(body.patterns).toMatchSnapshot();
});
    
test('body constructor',
     () => {
	 const mockSm = jest.mock();
	 const body = new Body({
 stateMachine: mockSm,
				debug: true,
});
     });


test('body constructor',
     () => {
	 const mockSm = jest.mock();
	 const body = new Body({
 stateMachine: mockSm,
				debug: true,
});
     });
