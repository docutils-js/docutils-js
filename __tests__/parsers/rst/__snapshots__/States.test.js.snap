// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Body patterns 1`] = `
Object {
  "anonymous": "__( +|)",
  "blank": " *$",
  "bullet": "[-+*\\\\u2022\\\\u2023\\\\u2043]( +|$)",
  "doctest": ">>>( +|$)",
  "enumerator": "((\\\\(\\\\))|(\\\\))|(\\\\.))( +|$)",
  "explicit_markup": "\\\\.\\\\.( +|$)",
  "field_marker": ":(?![: ])([^:\\\\\\\\]|\\\\\\\\.|:(?!([ \`]|$)))*(?<! ):( +|$)",
  "grid_table_Top": /\\\\\\+-\\[-\\+\\]\\+-\\\\\\+ \\*\\$/,
  "grid_table_top": /\\\\\\+-\\[-\\+\\]\\+-\\\\\\+ \\*\\$/,
  "indent": " +",
  "line": "([!-/:-@[-\`{-~])\\\\1* *$",
  "line_block": "\\\\|( +|$)",
  "option_marker": "((-|\\\\+)[a-zA-Z0-9]( ?([a-zA-Z][a-zA-Z0-9_-]*|<[^<>]+>))?|(--|/)[a-zA-Z0-9][a-zA-Z0-9_-]*([ =]([a-zA-Z][a-zA-Z0-9_-]*|<[^<>]+>))?)(, ((-|\\\\+)[a-zA-Z0-9]( ?([a-zA-Z][a-zA-Z0-9_-]*|<[^<>]+>))?|(--|/)[a-zA-Z0-9][a-zA-Z0-9_-]*([ =]([a-zA-Z][a-zA-Z0-9_-]*|<[^<>]+>))?))*(  +| ?$)",
  "simple_table_top": /=\\+\\( \\+=\\+\\)\\+ \\*\\$/,
  "text": "",
}
`;
